Issue/4399 notifications pending drafts

mzorz opened the pull request:
Fixes #4399

This implements a service that performs several checks on local drafts, in order to decide whether to hint the user on checking these drafts to finish / publish them.

![screenshot_20161126-163548](https://cloud.githubusercontent.com/assets/6597771/20643277/9a88c350-b3ff-11e6-868f-f421818782ab.png)


*(note the "0 days" in that screenshot was just for testing purposes - you'd always see a value different from zero and between 1 and 30 in there)* 👍 

#### Service flow

The service acts as follows:
1. checks all local drafts, and checks which have been pending for more than 3 days.
2. makes a differentiated notification for ONE draft pending, and another one for more than one pending draft
*ONE*: `One of your drafts has been pending for publishing for %d days. Would you like to check it?`
*ONE (generic)*: if the single pending draft has been sitting there for more than 30 days, or if we can't know for how long (for instance for a migrated draft from a previous version), we show this: `You\'ve got one draft pending for publishing. Would you like to check it?`
*MORE*: `You’ve got 2 posts in drafts. Want to check them?`
3. Notifications have quick actions: EDIT and IGNORE.
Tapping on a notification should take you to EDIT the draft, or the DRAFTS list (if there are drafts posts, it will take you there first over deciding to take you to the Pages list).
4. if the app had already a saved draft at the time this new version of the app is installed, and you update the app, we won't know when it was the time that draft was created, so we have a generic message for this case `You’ve got drafts pending for publishing. Would you like to check them out?`
5. don’t show any new notifications for the next 30 minutes if a notification about a specific draft has been already shown and `ignore` was tapped (or the notification was dismissed) in that timeframe.

#### Analytics tracking

3 events are being tracked:
- `NOTIFICATION_PENDING_DRAFTS_TAPPED` (the user taps on the notif)
- `NOTIFICATION_PENDING_DRAFTS_IGNORED` (the user taps on `Ignore` quick action)
- `NOTIFICATION_PENDING_DRAFTS_DISMISSED` (the user dismisses the notification )

#### Test cases
1) Existing drafts.
have the latest version of the app and make a draft post/page. Then update to this version, the database should be migrated but we can’t determine the date when the draft is created. Therefore, if the draft has been sitting there for more than 3 days (change the handset clock to test this), then a generic notification is shown:
```You\'ve got one draft pending for publishing. Would you like to check it?```

2) New draft. 
With this version of the app, create a drat post/page. Change the handset clock for more than 3 days in the future. Now open the app again and a notification should appear with the following text:
```One of your drafts has been pending for publishing for %d days. Would you like to check it?```, where `%d` is replaced with as many days in the future as your handset’s clock has been set to.

Also, if the draft has been pending for more than 30 days, the generic message is shown:
```You\'ve got one draft pending for publishing. Would you like to check it?```

3) More than 1 pending draft.
After doing as indicated in point 2, create another draft. The notification should change to the following text:
```You\'ve got %d posts in drafts. Would you like to check them?```, with %d set to the draft count.

Also, all these cases should work independently as you switch blogs/sites. Repeat for different blogs.

**Note** this feature only works with local drafts, so if you have other drafts on other devices / the web, this feature will not be aware of those.

#### QUICK ACTIONS & DISMISS:
- if you tap EDIT> this opens the draft editing view (or the post list if there's more than 1 post, or the Pages list if it's all pages pending)
- if you tap IGNORE: this dismisses the notification and doesn’t disturb you with this draft anymore
- if you dismiss the notification (by swiping left/right on it): it will be recreated again, next time the pending drafts are checked or when you switch the blog/site (and there are local drafts that match the conditions described above).

#### Tips for testing
In `NotificationsPendingDraftsService.java`:
- change line 88 to 
```                    long three_days_ago = now - (30000);   ```
so it’s 30 seconds instead of 3 days

- change line 29 to
```    private static final long MINIMUM_ELAPSED_TIME_BEFORE_REPEATING_NOTIFICATION = 1 * 60 * 1000;```
so it’s only 1 minute instead of 30 minutes.


#### For further thought
- the service in this PR is not started unless the user opens the app. Wondering if we should make this service be triggered in the background even if the user doesn't open the app that much, thus trying to get those users with "forgotten" pending drafts in there make it open more frequently.
- IDEA: design an actual NOTIFICATION block in the NOTIFICATIONS list for this, like the ones we have for achievements, with some sort of tip, like ```You’ve got a pending draft. Publishing your drafts will drive traffic to your site. Don’t let them stay there for too long!```

daniloercoli commented:
> private static final long MINIMUM_ELAPSED_TIME_BEFORE_REPEATING_NOTIFICATION = 1 * 60 * 1000;

Can we change this to ONE day? (It's even better if we change the value to samething > of ONE day). 
If I open the app many times during the day I will get notified every time I open it. 

IMO it's just very very annoying, making me to delete the draft altogether instead of publishing it.

daniloercoli commented:
> don’t show any new notifications for the next 30 minutes if a notification about a specific draft has been already shown and ignore was tapped (or the notification was dismissed) in that timeframe.

Can we intercept the swipe over the notification, and consider the swipe gesture as ignore?

mzorz commented:
> > private static final long MINIMUM_ELAPSED_TIME_BEFORE_REPEATING_NOTIFICATION = 1 * 60 * 1000;

> Can we change this to ONE day? (It's even better if we change the value to samething > of ONE day).
If I open the app many times during the day I will get notified every time I open it.

> IMO it's just very very annoying, making me to delete the draft altogether instead of publishing it.

Sure, a day should be fine. Added this in this commit 876e898

> Can we intercept the swipe over the notification, and consider the swipe gesture as ignore?

Oh it already has the same effect as tapping on Ignore (except it's not tracked in analytics), no need to intercept. What I meant to say is that ignoring or discarding the notification just doesn't do anything - the notification will appear again when the user brings the app to the foreground and there are still pending drafts older than 3 days and more than 30 minutes (now changing to 1 day) have elapsed since last time they opened the app.

mzorz commented:
Adding a comment from @nbradbury here:

> The notification asks a question ("would you like to check it?") but the actions aren't responses to that question. How about simply, "You drafted 'Post title here' three days ago but never published it." Tapping the notification would open the post for editing and swiping would dismiss it, so no actions are needed. It also might be nice to have a separate notification icon - perhaps the wp logo overlaid with a pencil?

Regarding the text and actions I was intending the notification to be hinting the user the idea that there *is something they can do* to change that informed status :). I somehow feel that  only informing the user about a situation,  without providing any obvious actions might make it feel less "actionable"  and more "informative", and what I'd ultimately want with this feature is the user to make a decision and take action. Either "Oh yes my post! Let's see what I'm missing"  or  "maybe later", that kind of thing :).

I like the idea about the icon - @drw158 what do you think (not only about the icon but the wording too please?)

daniloercoli commented:
> Oh it already has the same effect as tapping on Ignore (except it's not tracked in analytics), no need to intercept. What I meant to say is that ignoring or discarding the notification just doesn't do anything - the notification will appear again when the user brings the app to the foreground and there are still pending drafts older than 3 days and more than 30 minutes (now changing to 1 day) have elapsed since last time they opened the app.

I proposed to change the interval to one day since it seemed good enough with me few hours ago, but the more i think about this feature, the more it seems to me that we're going on a slippery road, and one day interval doesn't seem good now :)
Notify the users if they have pending drafts could be fine, but notify them each single day they open the app seems wrong to me. This way a user can't have a draft on their sites, without being notify at startup. 
Curious to hear what @nbradbury and @koke think about this.

Also, is there a setting in the app to shutdown this kind of notifications?

mzorz commented:
> Notify the users if they have pending drafts could be fine, but notify them each single day they open the app seems wrong to me. This way a user can't have a draft on their sites, without being notify at startup.

What if we actually set the IGNORE option to remember this, so the user doesn't get notified anymore for the specific drafts that triggered the notif? They will get notified again if any other drafts happen to fall into the same situation in the time to follow though.

koke commented:
Disclaimer: On iOS we don't (intentionally) support local drafts, so I'm thinking about server saved drafts, not sure how "local" changes things.

I have two drafts on my primary blog, both are over a year old. One of them, I don't think I'll ever publish, the other is a stub that I want to extend but I haven't organized my thoughts on it yet. Personally, I'd be annoyed at a notification like this. I'd be angry at a *repeating* notification for an abandoned draft, and I'd consider whether to delete the draft or the app (if it was any other app 😉). That's me though.

I think the intention is good, but I would ask: what is the goal here? If I were to guess, I'd say this is to help users that might want to post more often but lack motivation and could use a little nudge once in a while. This sounds great for those users, but they are just a subset, and I can think of many other cases where the feature is more annoying than useful: you have the motivation but not the time, you only write on weekends, you use drafts to collect ideas that might or might not end up published, you do lack the motivation but all the notifications achieve is increased anxiety and feeling of failure,... 

I'd be very careful in general when adding notifications that are not generated by an user event.

mzorz commented:
Of course, I see the point. 👍  I think a dialog saying *"Don't ask me about stuff like this again YES/NO"*  after you tap IGNORE on the quick actions for 2-3 times would cover all those cases where people actively intend to keep stuff on drafts and not ever publish them, just as you described.

Anyway, in our particular case here w/ local drafts, it's very likely the "subset" of people who should really be publishing / at least do something about their drafts is going to be greater in Android than on the web. 

> I'd be very careful in general when adding notifications that are not generated by an user event.

Agreed 100%.

Just a reminder, that these notifications are only triggered when you bring the app to the foreground, so you're also likely to be in a mood /  context where you can relate to the substance of the reminding notification itself, and at the same time it doesn't break your navigation loop (it doesn't get in the way of whatever you wanted to do).

drw158 commented:
https://github.com/wordpress-mobile/WordPress-Android/pull/4835#issuecomment-263272328

I agree that it's a little bit strange that it asks a question without yes/no for for action. I think it should be reworded.

I also think that action buttons would lead to better clarity and more posts published. The simplicity is enticing (no action buttons, just swiping), but I think button will be better overall for UX.

Maybe a middle ground? I think this wording is great:

> You drafted 'Post title here' three days ago but never published it.

with these buttons:

Edit / Ignore

No need for a new icon. Also we need to use Gridicons for any icons. 

https://github.com/Automattic/gridicons

daniloercoli commented on WordPress/src/main/java/org/wordpress/android/ui/notifications/services/NotificationsPendingDraftsService.java:
```diff
@@ -0,0 +1,229 @@
+    public static final String POST_ID_EXTRA = "postId";
+    public static final String IS_PAGE_EXTRA = "isPage";
+    private static final long MINIMUM_ELAPSED_TIME_BEFORE_REPEATING_NOTIFICATION = 24 * 60 * 60 * 1000; //a full 24 hours day
+    private static final long MAX_DAYS_TO_COUNT_DAYS = 30; //30 days
```
What do you think on renaming this constant to something like `MAX_DAYS_TO_SHOW_DAYS_IN_MESSAGE` ?
I know it's not that better, but at least there is the purpose of the constant in its name.

daniloercoli commented on WordPress/src/main/java/org/wordpress/android/ui/notifications/services/NotificationsPendingDraftsService.java:
```diff
@@ -0,0 +1,229 @@
+                        boolean isPage = post.isPage();
+
+                        //only notify the user if the last time they have been notified about this particular post was at least
+                        //30 minutes ago, but let's not do it constantly each time the app comes to the foreground
```
Comment is outdated. It refers to 30 minutes instead we changed it to 24h. Better to put the constant name there?

daniloercoli commented on WordPress/src/main/java/org/wordpress/android/ui/notifications/services/NotificationsPendingDraftsService.java:
```diff
@@ -0,0 +1,229 @@
+
+                        //only notify the user if the last time they have been notified about this particular post was at least
+                        //30 minutes ago, but let's not do it constantly each time the app comes to the foreground
+                        if ((now - post.getDateLastNotified()) > MINIMUM_ELAPSED_TIME_BEFORE_REPEATING_NOTIFICATION) {
```
This `if` condition can be moved up, immediately below ` Post post = draftPostsOlderThan3Days.get(0);`

daniloercoli commented:
I think this was done on purpose, but when I start the app and have local drafts on the device, no sound is played, and there is no new notification animation. Just the WordPress icon silently added to the status bar.

daniloercoli commented:
Just found a minor bug.
- Open the app and receive the "local drafts" notification.
- Go to posts, remove the draft from the list.
- Put the app in background.
- Open the notification and tap on it.
- The app is started and an error message is shown on the screen.

daniloercoli commented on WordPress/src/main/java/org/wordpress/android/ui/notifications/services/NotificationsPendingDraftsService.java:
```diff
@@ -0,0 +1,229 @@
+                        }
+                    }
+
+                    completed();
```
This should be move outside the `if` condition `if (draftPosts != null && draftPosts.size() > 0)`

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/notifications/services/NotificationsPendingDraftsService.java:
```diff
@@ -0,0 +1,235 @@
+    public static final int PENDING_DRAFTS_NOTIFICATION_ID = GENERIC_LOCAL_NOTIFICATION_ID + 1;
+    public static final String POST_ID_EXTRA = "postId";
+    public static final String IS_PAGE_EXTRA = "isPage";
+    //FIXME change this below line
```
Woops? https://bia.is/git-pre-commit-hook-avoid-committing-fixmes/

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/notifications/services/NotificationsPendingDraftsService.java:
```diff
@@ -0,0 +1,235 @@
+                if (draftPosts != null && draftPosts.size() > 0) {
+                    //now check those that have been sitting there for more than 3 days now.
+                    long now = System.currentTimeMillis();
+                    //FIXME change this below line
```
Woops number 2 ;) ?

mzorz commented:
All of the issues up here (including design and code) addressed now, though I still want to code something so tapping on `IGNORE` will actually remember the decision for this post, etc. so it's not too invasive. Will do in the following days.

Re this:
> I think this was done on purpose, but when I start the app and have local drafts on the device, no sound is played, and there is no new notification animation. Just the WordPress icon silently added to the status bar.

Yes it’s on purpose as to avoid being too annoying - we could maybe add some sound though? Or even a Dialog the first time only, or something like that? What do you think @daniloercoli?

daniloercoli commented:
> Yes it’s on purpose as to avoid being too annoying - we could maybe add some sound though? Or even a Dialog the first time only, or something like that? What do you think @daniloercoli?

I'm ok with it, Mario.
Can we mark this as not-ready-to-be-merged if you want to add other thing to it?

mzorz commented:
This is ready for another round, addressing:
- wording
- gridicon
- remembering if user taps on the `IGNORE` quick action for a specific draft ID or grouped notification (thus never bothering the user again on these specific drafts)
- user can disable/enable this altogether by switching the "Notify me on pending drafts" setting in "App Notifications" in My Profile tab. _Please note this is set to true by default_
- Added analytics tracking for that new switch each time a user enables/disables it.
- also addressed the bugs / comments on code made here so far

Screenshot for the notification settings screen here:

![screenshot_20161208-121413](https://cloud.githubusercontent.com/assets/6597771/21016805/c4c2aba0-bd45-11e6-8156-3c5cad9adc7a.png)

daniloercoli commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/AppPrefs.java:
```diff
@@ -446,4 +455,95 @@ public static void setReaderSwipeToNavigateShown(boolean alreadyShown) {
+        ArrayList<Long> idList = new ArrayList<>();
+        String idListString = getString(DeletablePrefKey.PENDING_DRAFTS_NOTIFICATION_IGNORE_LIST, "");
+        if (!TextUtils.isEmpty(idListString)) {
+            List<String> items = Arrays.asList(idListString.split("\\s*,\\s*"));
```
Can't this be simplified in ` List<String> items = Arrays.asList(idListString.split(","));` ?

mzorz replied:
addressed in 8e0bf59

daniloercoli commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/AppPrefs.java:
```diff
@@ -446,4 +455,95 @@ public static void setReaderSwipeToNavigateShown(boolean alreadyShown) {
+        if (!TextUtils.isEmpty(idListString)) {
+            // note wrapping the Arrays.asList call with a new object is needed because otherwise
+            // trying to remove an item from a List returned by Arrays.asList throws an UnsupportedOperationException
+            List<String> items = new ArrayList(Arrays.asList(idListString.split("\\s*,\\s*")));
```
And this other in `List<String> items = Arrays.asList(idListString.split(","));` ?

mzorz replied:
addressed in 8e0bf59

daniloercoli commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/AppPrefs.java:
```diff
@@ -15,8 +15,13 @@
+
 public class AppPrefs {
     private static final int THEME_IMAGE_SIZE_WIDTH_DEFAULT = 400;
+    private static final int MAX_PENDING_DRAFTS_AMOUNT = 100;
```
Maybe I'm not getting the point here, but wonder why there is a limit on the number of draft IDs stored in prefs.

mzorz replied:
Well the only place where we are deleting ids from the ignore list pref is when the draft is trashed/deleted. We should also delete ids (if found) when a draft is published (we aren't atm, will make the call in a next commit). But there's also the possibility case where for some reason ids can't be deleted one way or the other (let's call it random failure) and ids can be kept in that list forever (well, at least as long as you don't switch sites, or when you are an admin for one site only). I wouldn't want the ignore id list to grow indefinitely, so that's why this precautious method is there. A comment on that, 100 seemed a large enough number for this functionality, we could lower it to 15 or 20 and it'll pretty much be the same I think.

daniloercoli commented:
:shipit:

mzorz commented:
Notes for testing:

**CASE 1: more than 3 days, less than 30 days:**
- write a draft and hit back so it gets saved
- go to your device settings, set the date to some point in the future, beyond 3 days (i.e. for today Dec 19th, I set it to 23rd)
- bring the WordPress beta app back to the foreground, which starts the service described in this PR
- the notification should come up.

<img width="180" alt="screen shot 2016-12-19 at 5 43 46 pm" src="https://cloud.githubusercontent.com/assets/6597771/21328460/cb0e754a-c612-11e6-92f3-938f07aa4905.png">


**CASE 2: more than 30 days:**
- write a draft and hit back so it gets saved
- go to your device settings, set the date to some point in the future, beyond 30 days (i.e. for today Dec 19th, I set it to Jan 22nd)
- bring the WordPress beta app back to the foreground, which starts the service described in this PR
- the notification should come up.

if the draft has been pending for more than 30 days, the generic message is shown:
`You drafted '<post title>' but never published it.`

**CASE 3. More than 1 pending draft.**
After doing as indicated in point 1, create another draft, get the app to the background, and get it again on the foreground. The notification should change to the following text:

`You drafted <number of draft posts> posts but never published them. Tap to check.`

Also, all these cases should work independently as you switch blogs/sites. Repeat for different blogs.

Note this feature only works with local drafts, so if you have other drafts on other devices / the web, this feature will not be aware of those.

**QUICK ACTIONS & DISMISS:**

- if you tap EDIT> this opens the draft editing view (or the post list if there's more than 1 post, or the Pages list if it's all pages pending)
- if you tap IGNORE: this dismisses the notification and doesn’t disturb you with this draft anymore
- if you dismiss the notification (by swiping left/right on it): it will be recreated again, next time the pending drafts are checked (after 24 hours) or when you switch the blog/site (and there are local drafts that match the conditions described above).

For swipe dismissal, you can check that the notification is re-created 24 hours by setting the device date to the next day after having dismissed the notification. On the contrary, tapping IGNORE will make this notification go away for good for this draft.


cc @rachelmcr

rachelmcr commented:
Thanks, @mzorz. I tried these steps on my device (using the latest alpha release, alpha-31) and I still can't trigger the notification. Is that because I'm on the alpha version instead of a beta or release version?

mzorz commented:
**EDIT**: just updated the comment above which now includes title `DISABLE PENDING DRAFTS NOTIFICATIONS`
